src/main_thread/init/utils/update_manifest_codec_support.ts

```diff
@@ -0,0 +1,155 @@
+/**
+ * Ensure that all `Representation` and `Adaptation` have a known status
+ * for their codec support and probe it for cases where that's not the
+ * case.
+ *
+ * Because probing for codec support is always synchronous in the main thread,
+ * calling this function ensures that support is now known.
+ *
+ * @param {Object} manifest
+ */
+export function updateManifestCodecSupport(
+  manifest: IManifestMetadata,
+): ICodecSupportList {
+  const codecSupportList: ICodecSupportList = [];
+  const codecSupportedByMseMap: Map<string, Map<string, boolean>> = new Map();
+  const codecSupportedByCdmMap: Map<string, Map<string, boolean>> = new Map();
+
+  manifest.periods.forEach((p) => {
+    [
+      ...(p.adaptations.audio ?? []),
+      ...(p.adaptations.video ?? []),
+      ...(p.adaptations.text ?? []),
+    ].forEach((a) => {
+      let hasSupportedCodecs = false;
+      a.representations.forEach((r) => {
+        const isEncrypted = r.contentProtections !== undefined;
+        if (r.isSupported !== undefined) {
+          if (!hasSupportedCodecs && r.isSupported) {
+            hasSupportedCodecs = true;
+          }
+          return;
+        }
+        let isSupported = false;
+        const mimeType = r.mimeType ?? "";
+        let codecs = r.codecs ?? [];
+        if (codecs.length === 0) {
+          codecs = [""];
+        }
+        for (const codec of codecs) {
+          const isSupportedByMSE = isCodecSupportedByMSE(mimeType, codec);
+          // if MSE supports the codec, and the content is encrypted,
+          // check further if the CDM also supports the codec.
+          if (isSupportedByMSE.supported && isEncrypted) {
+            const isSupportedByCDM = isCodecSupportedByCDM(mimeType, codec);
+            isSupported = isSupportedByMSE.supported && isSupportedByCDM;
+          } else {
+            isSupported = isSupportedByMSE.supported;
+          }
+
+          if (!isSupportedByMSE.wasKnown) {
+            codecSupportList.push({
+              mimeType,
+              codec,
+              result: isSupportedByMSE.supported,
+            });
+          }
+
+          if (isSupported) {
+            r.codecs = [codec];
+            break;
+          }
+        }
+
+        r.isSupported = isSupported;
+        if (r.isSupported) {
+          hasSupportedCodecs = true;
+        }
+        if (!hasSupportedCodecs) {
+          if (a.isSupported !== false) {
+            a.isSupported = false;
+          }
+        } else {
+          a.isSupported = true;
+        }
+      });
+    });
+    ["audio" as const, "video" as const].forEach((ttype: ITrackType) => {
+      const forType = p.adaptations[ttype];
+      if (forType !== undefined && forType.every((a) => a.isSupported === false)) {
+        throw new MediaError(
+          "MANIFEST_INCOMPATIBLE_CODECS_ERROR",
+          "No supported " + ttype + " adaptations",
+          { tracks: undefined },
+        );
+      }
+    });
+  });
+  return codecSupportList;
```

Comment:
So if I followed, this function will update the `isSupported` property based on both MSE and EME checks but only return the list based on the MSE check?

Why not both, identified in some ways as we have the information.